### DISCUSSION TITLE
Running Optuna on remote Dask cluster (on K8)

### DISCUSSION THREAD

**Discussion by raz-gilad** [optuna/optuna#4731] [Q&A] [answered]
Question on : optuna v3.1.1
I'm using Optuna with remote Dask storage integration (SQLite on the dask schedular pod) .
I'm using TPESampler with constant_liar=True
The Study has n_startup_trials=5

As each trial take a few hours (heavy NN) I'm sending the first 5 trials in parallel:
_futures = [client.submit (study.optimize, func, n_trials=1,n_jobs=1, pure=False) for _ in range(5)]_
Problem is, the **suggested parameters are the same** for each of the trials.
Any idea how to fix this ?

note - 
I can use _client.submit (study.optimize, func, n_trials=5,n_jobs=5, pure=False)_ but the K8 dask worker pod cannot be scalled vertically (no more than 1.75 cores) , so it will not really run in parallel any how.
Ideas ?

**Comment by nzw0301** [comment]
When I ran the following code inspired by your question and Optuna's dask example: https://github.com/optuna/optuna-examples/blob/main/dask/dask_simple.py, the suggest params were different. 


```python
import optuna

from dask.distributed import Client
from dask.distributed import wait
from sklearn.datasets import load_digits
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split


def objective(trial):
    X, y = load_digits(return_X_y=True)
    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2)

    max_depth = trial.suggest_int("max_depth", 2, 10)
    n_estimators = trial.suggest_int("n_estimators", 1, 100)

    clf = RandomForestClassifier(max_depth=max_depth, n_estimators=n_estimators)
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    score = accuracy_score(y_test, y_pred)
    return score


if __name__ == "__main__":
    with Client(processes=False) as client:
        print(f"Dask dashboard is available at {client.dashboard_link}")
        storage = optuna.integration.dask.DaskStorage()
        study = optuna.create_study(storage=storage, direction="maximize",
                                    sampler=optuna.samplers.TPESampler(seed=1, constant_liar=True, n_startup_trials=5))
        # Submit 10 different optimization tasks, where each task runs 7 optimization trials
        # for a total of 70 trials in all
        futures = [
            client.submit(study.optimize, objective, n_trials=1, pure=False, n_jobs=1) for _ in range(5)
        ]
        wait(futures)
        print(f"Best params: {study.best_params}")
```

Note that the main dependencies are Optuna==3.1.1 and dask==2023.5.1.

Could you share more details of the code?

**Reply by raz-gilad** [reply]
Thanks for the reply.

I think the major differences are:

1. My Dask is remote :
	_client = Client (url)
	client.restart(wait_for_workers=True)_
	
2. I use sqllite as the coordination point:
	_backend_storage = optuna.storages.RDBStorage("sqlite:///dask_optuna.db")
	dask_storage = optuna.integration.DaskStorage (storage=backend_storage, client=client)_	
	
Best guess is that the serialization process (in dask submit to the remote cluster)  eventually creats same sampler (TPE) on each dask worker (different machines) . Also, rememebr that the process of suggesting values is random at fisrt (it takes a few hours until I get the first results) and it uses same seed.
Maybe I need to recreate sampler / change sampler seed before each trial .

**Reply by raz-gilad** [reply]
One more - I found out that if I use n_trials=2 (or more) for the startup trials , (sending more than 1 trial per dask task), it works OK. 
It could also be that when n_trials=1 in remote mode, the sampler constant_liar is not cheked.

**Reply by nzw0301** [reply]
Thank you for your reply 

> Maybe I need to recreate sampler / change sampler seed before each trial.

I agree with you. I think not specifying `seed` of the sampler argument is a simpler solution to avoid using the same suggestions by Optuna.

**Reply by raz-gilad** [reply]
10x, not specifying seed will indeed lead to random seed, but it will not change during the runs.

after some digging , I found the follwing in the BaseSampler (_base.py) class with regards to 'reseed' function:

        """Reseed sampler's random number generator.

        This method is called by the :class:`~optuna.study.Study` instance **if trials are executed
        in parallel with the option `n_jobs>1`. In that case, the sampler instance will be
        replicated including the state of the random number generator, and they may suggest the
        same values. To prevent this issue, this method assigns a different seed to each random
        number generator.
        """
Yo can see also in the _optimize.py code that :
```
 if n_jobs == 1:
            _optimize_sequential(..., reseed_sampler_rng=False,, ....)
```

Do you think it is a bug in code ?

**Reply by nzw0301** [reply]
TBH, I don't know because I'm unfamiliar with dask, sorry.

**Reply by raz-gilad** [reply]
Thanks for the attention nzw0301.
Anyway,to summerize :

-   study.sampler.reseed(... per trial  - did not work.
- study.sample = TPESampler(... per trial  - did not work.
- create new study per trial - worked 

so the code for unique startup runs is eventauly :

```
	runs=[]
	for _ in range(num_startup_trials):
		runs.append( client.submit (study.optimize, func, n_trials=1,n_jobs=1, pure=False))
		study = optuna.create_study (storage=dask_storage,
		                             sampler=TPESampler (constant_liar=True, n_startup_trials=1),
		                             study_name=study_name, direction="maximize", load_if_exists=True)

	wait(runs)
```

**Reply by raz-gilad** [reply]
and it does look like an open end in optuna that needs fixing.

**Reply by nzw0301** [reply]
Indeed, I realised that https://github.com/optuna/optuna/issues/4604 reported a similar issue.

**Comment by raz-gilad** [comment, marked as answer]
Found a simple workaround  for making the parameters of the startup trials differnt.

The problem :
When sending the tasks to DASK we loop on 'study.optimize' with "n_trials=1" , to make one DASK worker per each  trial.
This cause the startup trials to have the same random startup parameters, as with 1 trial Optuna does not check for other runs.

The solution :
Override the number of **jobs** to be more than 1 - that's all
In this case , Optuna will know it is a parallel env and will check the storage for the other trials.

`startup_runs = [client.submit(study.optimize, objective_function, n_trials=1, n_jobs= num_stratup_trials, pure=False) for _ in range(num_stratup_trials) ]
wait(startup_runs)`
